### ISSUE TITLE
Remove require-jsdoc

### ISSUE THREAD

**Issue by desandro** [fictivekin/eslint-config-fk#5]
[End-of-Life for Built-in JSDoc Support in ESLint](https://eslint.org/blog/2018/11/jsdoc-end-of-life/)

> The require-jsdoc and valid-jsdoc rules will be [deprecated](https://eslint.org/docs/user-guide/rule-deprecation#rule-deprecation). These two rules will remain in ESLint but we will no longer add new features or fix bugs for them. These rules may be removed in a future major release of ESLint.

I recommend removing `require-jsdoc` all together. I find it more a nuisance than useful. If we do want JS Doc stuff, then we can add it back with [eslint-plugin-jsdoc](https://github.com/gajus/eslint-plugin-jsdoc)

**Comment by jakefentress** [comment]
Totally understand - this rule annoys me more than any other. But, it annoys me to get me to do some documentation I _should_ have done to begin with. Short of going to typescript, I find it helpful to document parameters and the types of data they are expecting.

Do you have any thoughts on how we might approach documentation and/or encourage the team towards documentation without this?

**Comment by BrianaCamp** [comment]
TEAM KEEEEEP

**Comment by lea37** [comment]
Team keep IF we can customise the rules. Please show me how to turn off some stuff like return 😅

**Comment by nerdpruitt** [comment]
I like having something that pesters me to add documentation, but I would like to relax the rules. For example, sometimes param or return descriptions can be redundant and not all that useful, and you can turn those off with `requireParamDescription` and `requireReturnDescription`

**Comment by printerscanner** [comment]
I personally don't like the required parameters part of this. 

1. I have trouble remembering the syntax, and 
3. I don't find it to be the most straightforward way to describe a function's purpose. And then a comment can end up having enough technical jargon in it that it's easier to just read the code (because at least it's colorful in the code). 

But, I agree with Sandy that if it's a contentious issue, that means we should keep it.

**Comment by jakefentress** [comment]
per @nerdpruitt:

> I like having something that pesters me to add documentation

I agree. Interestingly, having strict rules here doesn't mean that you add accurate or meaningful documentation. Sometimes a simple "psst, don't forget to document this for future you and your teammates" is probably the best thing we're trying to achieve here.

For me, knowing the type of data expected is helpful. I guess that's why typescript has really taken off, but I don't want that level of forced conformity, so forcing it here feels maybe like the wrong thing.

I don't want to use Ashur's words in a way he might not have intended, but I do think having a team culture of documenting _is_ a good thing. I've been too lax there in the past — just trying to move fast, thinking how I write things is plenty self-documenting, deadlines, etc — and we can see how that has made it harder for people to get up to speed.

**Comment by ashur** [comment]
**tl;dr** — I think tooling that enforces _some_ kind of documentation is important as part of our collective code building; whether that's `require-jsdoc`, or something else that helps us maintain what I see as benefits of documentation, I'm game.

> **Note** — This is primarily in reference to non-TypeScript projects ☺️

## Benefits of Documentation

Things I really appreciate from documented code:

- **Showing my work** — it's an inline artifact of what I was thinking at the time I wrote that code, and how I intended it to be used
- **Increased speed of understanding** — it's slower the first time to document, sure, but the _next_ time I come back I can just read the [CliffsNotes](https://en.wikipedia.org/wiki/CliffsNotes) instead of the whole book
- **Increased speed of usage** — I get TypeScript-like autocomplete of those functions and their parameters in my editor, which speeds me up during development

JSDoc is a little goofy and maybe kind of old, but it has a [nice, cross-language history](https://en.wikipedia.org/wiki/Docblock) and lots of [native support](https://code.visualstudio.com/Docs/languages/javascript#_jsdoc-support) in tooling.

## In Practice

JSDoc supports a _ton_ of properties, but my ideal block looks something like this:

```javascript
/**
 * Shortest possible summary of what the function does
 * @param {type} name - optional summary where warranted; use your best judgment
 * @return {type}
 */
```

### Summary

```javascript
/**
 * Shortest possible summary of what the function does
```
Helps me understand what the function does without having to read and interpret the code. Sometimes I'm just perusing things rather than using them (reading utility functions, filters, shortcodes, etc.) just to familiarize myself with a codebase.

This can feel redundant for very simple functions. I definitely get that! My personal opinion is to trust your best judgment — can the next developer who looks at your code:

- PR reviewers
- new folks who are onboarding
- you in six months 😅

understand what it does just by reading the function name and its signature?

```javascript
function addTwoNumbers(a, b) {
```

If so, then maybe a summary is OK to skip 👍

### Parameters

> **Note** — This is an important part of populating editor autocompletion

```javascript
@param {type} name [- optional summary]
```

In short, what does this function expect me to pass along when I call it?

Data types are particularly important to me here: I can read through the code to figure out that `foo` could be a string, or an Object, or an instance of `Foo`, but it's a lot faster and more reliable to see it documented all in one spot:

```javascript
@param {string|Object|Foo} foo
```

(In many, many cases, the variable name is self-documenting. In that case, I think it makes a lot of sense to skip the summary!)


### Return Values

> **Note** — This also helps populate editor autocompletion

```javascript
 * @return {type}
 */
```

What should I expect when I call this function? An array? A `Promise` I need to `await`? Nothing at all?

Again, this is something you _can_ glean from reading over the code, but (in my opinion) relying on that exclusively would tend to _decelerate_ our overall velocity rather than accelerate it.

## Personal Notes

I have two snippets in my text editor to make building out a JSDoc block quick and easy:

`/**` →

```
/**
 * @param {type} name
 * @return {type}
 */
```

and `@param` →

```
@param {type} name
```

to quickly add additional `@param` entries.

**Comment by desandro** [comment]
TODO Sandy - add writing JSDocs to the UXE Styleguide
